Makefile

```diff
@@ -1,22 +1,180 @@
-DIR := ${CURDIR}
-build_dir := $(DIR)/.build
+.PHONY: default all help
+
+default: all
+
+all: lint test
+
+help:
+	@echo "Usage: make <target>"
+	@echo
+	@echo "  all              - lints and tests the code"
+	@echo "  lint             - lints the code"
+	@echo "  test             - tests the code"
+	@echo "  generate         - generate protocol buffer and gRPC stub code (default)"
+	@echo "  generate-check   - ensure generated code is up to date"
+
+# Used to force some rules to run every time
+FORCE: ;
+
+############################################################################
+# OS/ARCH detection
+############################################################################
+os1=$(shell uname -s)
+os2=
+ifeq ($(os1),Darwin)
+os1=darwin
+os2=osx
+else ifeq ($(os1),Linux)
+os1=linux
+os2=linux
+else
+$(error unsupported OS: $(os1))
+endif
+
+arch1=$(shell uname -m)
+ifeq ($(arch1),x86_64)
+arch2=amd64
+else ifeq ($(arch1),aarch64)
+arch2=arm64
+else
+$(error unsupported ARCH: $(arch1))
+endif
+
+#############################################################################
+# Vars
+#############################################################################
+
+build_dir := ${CURDIR}/.build/$(os1)-$(arch1)
+
+protoc_version = 3.14.0
+ifeq ($(arch1),aarch64)
+protoc_url = https://github.com/protocolbuffers/protobuf/releases/download/v$(protoc_version)/protoc-$(protoc_version)-$(os2)-aarch_64.zip
+else
+protoc_url = https://github.com/protocolbuffers/protobuf/releases/download/v$(protoc_version)/protoc-$(protoc_version)-$(os2)-$(arch1).zip
+endif
+protoc_dir = $(build_dir)/protoc/$(protoc_version)
+protoc_bin = $(protoc_dir)/bin/protoc
+
+protoc_gen_go_version := $(shell grep google.golang.org/protobuf v2/go.mod | awk '{print $$2}')
+protoc_gen_go_base_dir := $(build_dir)/protoc-gen-go
+protoc_gen_go_dir := $(protoc_gen_go_base_dir)/$(protoc_gen_go_version)-go$(go_version)
+protoc_gen_go_bin := $(protoc_gen_go_dir)/protoc-gen-go
+
+protoc_gen_go_grpc_version := v1.0.1
+protoc_gen_go_grpc_base_dir := $(build_dir)/protoc-gen-go-grpc
+protoc_gen_go_grpc_dir := $(protoc_gen_go_grpc_base_dir)/$(protoc_gen_go_grpc_version)-go$(go_version)
+protoc_gen_go_grpc_bin := $(protoc_gen_go_grpc_dir)/protoc-gen-go-grpc
+
 golangci_lint_version = v1.24.0
 golangci_lint_dir = $(build_dir)/golangci_lint/$(golangci_lint_version)
 golangci_lint_bin = $(golangci_lint_dir)/golangci-lint
 
+apiprotos := \
+	v2/proto/spiffe/workload/workload.proto \
+
+#############################################################################
+# Toolchain
+#############################################################################
+
+go_version_full := 1.13.15
```

Comment:
Do we want/need to document supported go version(s) somewhere else? I think this is the first time it is being codified somewhere?

Reply:
As a module, the canonical place is the go version in the go.mod?

Reply:
(since we don't produce an artifact)

Reply:
Ok fair ... The thing in my mind is that people often ask what the minimum supported go version is for a given lib. We can run the test tooling with a different version, so I don't know that what we put here is strictly the same as what is "supported", but I figured if we are building with 1.13 then maybe that's a good place to start.

Reply:
Yeah, i think ideally we test with each version between the supported version and the latest. Maybe we can get a CI/CD refresh on this repository with github actions that does that.

Reply:
Yep ... I went to file an issue for that and found that it already exists https://github.com/spiffe/go-spiffe/issues/30

Reply:
The [go directive](https://golang.org/ref/mod#go-mod-file-go) in go.mod can have some subtle side effects on dependency resolution as well as preventing compilation for using featurews newer than indicated in the directive.